### ISSUE TITLE
Feature request: passedProps()

### ISSUE THREAD

**Issue by xanf** [vuejs/test-utils#2151]
One thing which annoys me **a lot** is that 99% of time I want to assert that I'm **passing** something to underlying component in my component (my area of responsibility) but in order to write a test correctly I should be aware of implementation details of the component - I need to use `.props()` for everything declared as prop, `.attributes()` for everything that ends up on root DOM node, and sometimes even `vm.$attrs` if I'm either using attributes inheritance and/or `v-bind="$attrs"` in the code

At the moment at GitLab there are  >60 instances of `vm.$attrs` in tests and this is the source of infinite confusion. Additionally, if something one day migrates from `$attrs` to props - tests should be altered, although there are no changes to component functionality

I've created [StackBlitz](https://stackblitz.com/edit/vitejs-vite-whcwjh?file=src%2Fcomponents%2FIPassThreeProps.test.js) to demonstrate what I would like to have :)

Basically I'm proposing to have new method on `wrapper` (proposal: `.passedProps()` but I'm open to alternatives) which will basically return `{ ...this.$props, ...this.$attrs }` (whenever it should consider default prop values is open to discussion, but I'm ok what is easier to implement)

If this makes sense to you, I will be happy to open pull request with implementation and docs page with mini-guide, what is the difference between `.attributes()`, `.props()` and `.passedProps()`.

(In perfect world I would like to alter behaviour of `.attributes()` on Vue instances to return `$attrs`, but for reverse compatibility we're late to party)

**Comment by xanf** [comment]
@lmiller1990 @cexbrayat WDYT of this one?

**Comment by xanf** [comment]
Additional rationale: why do not have this one as plugin?

Because I really believe that asserting "what I am passing" should be the core functionality of the testing library and having it separated feels quite weird

**Comment by lmiller1990** [comment]
I personally don't see this as something I'd use. I much prefer to assert against the outputs, which is normally the DOM, but there's different philosophies here, we really should let people do whatever they want, so I see why this might be useful for some users. 

Interested to see if anyone else finds this useful, though, so happy to at least get a few more opinions. I don't really have any issues against including it, it sounds like it'll just be a thin layer on top of props / attrs.

The name would certainly need consideration. That's probably my main concern. I'd expect `passedProps()` to be the same as `props` - even if the lines are kind of blurred, I think the average developer would see the `passedProps` and think it will be anything they passed as props. Eg given `<div foo="bar" />`, `passedProps()` would be `{ foo: "bar" }`.

Maybe @cexbrayat or @freakzlike has some input / thoughts.

**Comment by xanf** [comment]
@lmiller1990 Thank you for your input. I agree that `passedProps()` wording might be brittle, and I totally agree that this has way-more value for shallow-mount-like adepts :) However asserting that the component is passing something correctly is extremely critical for big codebases 

However, I'm not sure about your example with `<div foo="bar" />`. Let's talk about something like `<my-cmp foo="bar" />`. I think we're on the same page that we will expect that `.passedProps()` will be `{ foo: "bar" }`, while now it might be either in `.props()` or `.attributes()` or even `.vm.$attrs` depending on **implementation detail** of `my-cmp`. Just trying to figure out your "passed as props" concern, because for me it seems we're literally on the same position

**Comment by lmiller1990** [comment]
I don't actually know the intricacies of props/attrs as well as some, if I do `foo="bar"` without declaring `props` in my component, does this mean `foo` comes in on `$attrs`?

I feel like many people do not understand this, mostly they just want know that a component received a value, which is basically your proposal.

I want to think of a name that represents "all the data the component got passed". `propsAndAttrs()` probably captures this best, but I don't think that's a great name. WDYT?

**Comment by lmiller1990** [comment]
Also, wouldn't it be `receivedProps`? Or maybe it depends how you interpret the wording...

**Comment by markrian** [comment]
To give another perspective on this, I think this would be analagous to asserting that a spied-on function was called with certain arguments.

```js
expect(spy).toHaveBeenCalledWith({ foo: 'bar', qux: { bool: true }, class: 'bold' });
```

How the real function _uses_ those arguments is irrelevant.

A stubbed component child of a shallow mounted component takes the role of the spied-on function. Only the real child component knows which are attrs and which are props. The parent doesn't care.

**Comment by cexbrayat** [comment]
This is not something I'm personally missing, so I don't think we need to add this to VTU APIs
As a plugin or a custom matcher maybe?

**Comment by lmiller1990** [comment]
@markrian are you for / against this feature? I'm unsure.

Test Utils feels complete to me, I would be happy if the API stayed the same. Is this possible using the plugin API? That might be a good way to see if our plugin API is appropriate or not.

**Comment by markrian** [comment]
@lmiller1990 I'm in favour of it. We're still using v1, which doesn't provide a plugin API as far as I can tell, so that's not an option for us. Upgrading to Vue 3 is a long process :)

But, to give another angle on @xanf's [point](https://github.com/vuejs/test-utils/issues/2151#issuecomment-1665923891) about this being fundamental, the main use case for us is testing components that are thin wrappers around other components (i.e., pass through all attrs). Something like this:

```
<script>
import { RealComponent } from 'a-library';

export default {
  inheritAttrs: false,
  components: { RealComponent },
};
<script>

<template>
  <real-component v-bind="$attrs" />
</template>
```

Though in reality they might present a slightly modified API, say, adding their own props on top and ignoring or overriding others. It could be argued that such components should be explicit about which props they use/pass through (perhaps via something like `props: RealComponent.props`), but that's not always practical.

As it is, `@vue/test-utils` doesn't have a way to assert what was passed as attrs/props to that component, hence this feature request. It does feel like a missing piece of the library. While the plugin API of v2 seems like it could be used to implement this, it seems to me that it should be possible to test that example component. (Or rather: be able to assert that component was given the expected attrs/props in some _other_ component's shallow mounted tests, which this component is a child of.)

**Comment by freakzlike** [comment]
I like the idea and I think it makes sense when testing with a component library. Not sure if it's worth to add it to VTU or instead create a Plugin

**Comment by lmiller1990** [comment]
@xanf any estimate on code complexity / difficulty to implement this?

**Comment by julianpung** [comment]
A little bit late to the party, but I think you (@xanf) should rethink your tests by looking at the basics of unit testing.

You do not want to make assertions on the internal implementation of your component, but on the actual feature that depends on that prop. That way your tests actually make sure that the functionality remains the same.

So instead of asserting `expect(target.vm.$attrs.two).toBe('bar');`, you should be looking at something like `expect(target.find('p').text()).toBe('Consumed prop one: bar');`

**Comment by doug-wade** [comment]
@lmiller1990 I implemented this as a plugin [here](https://github.com/doug-wade/vtu-plugin-received-props/blob/main/index.ts), and it seems very straightforward -- I'm not sure it even needs to be published to the registry?

**Comment by lmiller1990** [comment]
I am inclined to agree, this feature is useful to some specific use cases, but there is something to be said for keeping our API surface smaller and encouraging what is generally considered a good practice of not testing implementation details. I'd recommend anyone needing it to copy/paste those 5 lines into their project.
